Json schema

carlhiggs opened the pull request:
This pull request addresses #506, introducing 

1.  a JSON Schema definition (`region-json-schema.json`) against which we can validate our example configuration file (and later, user configuration files) based on the [2020-12 draft specification](https://json-schema.org/draft/2020-12)
2. a test that the example configuration matches the region json schema 

Based on my local interactive tests our current example configuration file appears to validate against the region JSON schema, however the inclusion of the test is to ensure that this remains the case in our continuous integration workflow.

There is much more to do, but this will ultimately involve breaking changes to region configuration files.  For example, consistent formatting of dates, and metadata for data objects.

But for now, we now have a draft JSON schema configuration file definition that validates, and this is something that we can build from.  This addresses #506 in a most basic way; however, let's leave that issue open until we introduce checking of user configurations.  

Checking of user configurations will require more rigorous thinking through of what are required vs optional fields.  Further, changes to configuration formatting requirements will cascade to code changes (hopefully, simplifications) to accommodate these.  So, this is all part of a bigger task.

carlhiggs committed:
updated schema to most recent reference standard (validated the updated schema against the reference standard, and the example configuration against the updated version successfully using jsonschema) towards #506

carlhiggs committed:
updated schema and tests.py to evaluate example configuration matching schema, towards #506

carlhiggs committed:
updated schema and tests.py to evaluate example configuration matching schema, towards #506; example configuration now validates (although, it is pending improvement and standardisation, eg of dates --- but that will involve breaking changes, so is a separate bigger task

carlhiggs committed:
updated tests.py with jsonschema test using correct validation conditions (i.e. that null is correct value, not 0); towards #506

carlhiggs commented:
there are still some issues... that's why there are tests...

carlhiggs committed:
trying again the tests workflow; there's some wierd interaction in the yaml constructuor modificatino for parsing dates and keys as strings; hopefully test passes in workflow now (does locally)

carlhiggs commented:
So, the test for valid yaml schema appears to check out fine 
```
test_0_2_schema_yaml (tests.tests.tests.test_0_2_schema_yaml)
Check if example configuration file is valid against jsonschema file. ... ok
```
----- but somehow unexpectedly, I think that the work around to ensure that yaml keys are treated as strings not integers although intended to only apply within the scope of the test function must be working more globally, as an error related to this occurs now in a subsequent test ---- expecting a string, but getting an int:

<details>

<summary>A couple of failing tests on the image list</summary>

```
ERROR: test_6_example_generate (tests.tests.tests.test_6_example_generate)

Generate resources for example region.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ghsci/process/tests/tests.py", line 132, in test_6_example_generate
    r.generate()
  File "/home/ghsci/process/subprocesses/ghsci.py", line 1028, in generate
    generate_resources(self)
  File "/home/ghsci/process/generate.py", line 114, in generate
    r.generate_report(language=language, report='indicators')
  File "/home/ghsci/process/subprocesses/ghsci.py", line 1080, in generate_report
    generate_report_for_language(
  File "/home/ghsci/process/subprocesses/_utils.py", line 211, in generate_report_for_language
    phrases = r.get_phrases(language)
              ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ghsci/process/subprocesses/ghsci.py", line 1842, in get_phrases
    phrases[f'Image {i} file'] = city_details['images'][i]['file']
                                 ~~~~~~~~~~~~~~~~~~~~~~^^^
KeyError: 5

======================================================================
ERROR: test_8_example_generate_report_in_another_language (tests.tests.tests.test_8_example_generate_report_in_another_language)
Generate resources for example region.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/ghsci/process/tests/tests.py", line 174, in test_8_example_generate_report_in_another_language
    r.generate_report('Spanish - Latin America')
  File "/home/ghsci/process/subprocesses/ghsci.py", line 1080, in generate_report
    generate_report_for_language(
  File "/home/ghsci/process/subprocesses/_utils.py", line 211, in generate_report_for_language
    phrases = r.get_phrases(language)
              ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ghsci/process/subprocesses/ghsci.py", line 1842, in get_phrases
    phrases[f'Image {i} file'] = city_details['images'][i]['file']
                                 ~~~~~~~~~~~~~~~~~~~~~~^^^
KeyError: 5
```

</details>


Not sure how to resolve this right now -- but that's the issue, can address tomorrow.  Must be some simple way to specify the parsing of keys only as strings only within the scope of the schema validation test.

carlhiggs committed:
reapproached tests.py with simplified jsonschema test; towards #506.  Test workflow appears to pass now, without unintended downstream consequences of modifying pyyaml constructors.

shiqin-liu commented:
Thanks for the work Carl! As this looks like a work-in-progress to me, let's keep it i the development branch for now, rather than merging into the main. This work deserve more thinkings and discussion, circle back to the goal of replacing with Json schema, I see there are opportunities to break down based on different functional inputs, and possibly to separate the input data parameters with meta data information etc.

carlhiggs commented:
Hi Shirley,

Yes, it's describing the current state -- it's future work to change
things. I was thinking of doing the changes incrementally, rather than all
at once,

Enjoy the weekend when it comes,

Carl

On Fri, 21 Mar 2025 at 09:14, Shirley Shiqin Liu ***@***.***>
wrote:

> Thanks for the work Carl! As this looks like a work-in-progress to me,
> let's keep it i the development branch for now, rather than merging into
> the main. This work deserve more thinkings and discussion, circle back to
> the goal of replacing with Json schema, I see there are opportunities to
> break down based on different functional inputs, and possibly to separate
> the input data parameters with meta data information etc.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/healthysustainablecities/global-indicators/pull/526#issuecomment-2741784109>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADDCCMXBN4TU3PLYIKWVAOL2VM4U5AVCNFSM6AAAAABZHHYLS2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBRG44DIMJQHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
> [image: shiqin-liu]*shiqin-liu* left a comment
> (healthysustainablecities/global-indicators#526)
> <https://github.com/healthysustainablecities/global-indicators/pull/526#issuecomment-2741784109>
>
> Thanks for the work Carl! As this looks like a work-in-progress to me,
> let's keep it i the development branch for now, rather than merging into
> the main. This work deserve more thinkings and discussion, circle back to
> the goal of replacing with Json schema, I see there are opportunities to
> break down based on different functional inputs, and possibly to separate
> the input data parameters with meta data information etc.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/healthysustainablecities/global-indicators/pull/526#issuecomment-2741784109>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADDCCMXBN4TU3PLYIKWVAOL2VM4U5AVCNFSM6AAAAABZHHYLS2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBRG44DIMJQHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>

carlhiggs commented:
Hi @shiqin-liu and looping in @gboeing and @rschifan,

Sorry I couldn't make the rescheduled software catchup on Friday; I'm just now back at work.  Not sure if you folk had a chance to discuss the configuration schema topic in your catch up --- it relates to @rschifan's review, although obviously it is only a small part of that bigger journey towards a software architecture refresh.  

This pull request relates to our previous discussions (#506) where we agreed that I would prepare a JSON Schema definition for our existing configuration file as a first step.    This is a non-breaking change, although I also introduced a test that validates adherence of the example configuration file to the existing schema; that is, its designed to pass given the current state of the configuration file.

If I add more to this existing pull request, the changes will soon become quite large and challenging to review, with changes to configuration soon necessitating downstream changes to code --- everything will snowball.  I feel like the existing pull request addresses the modest goal of #506: I implemented a JSON schema, and using the Python jsonschema library (added to our software environment in a previous pull request, in preparation for this one) I use the schema to validate our existing configuration file.

I propose that we merge this pull request to address this first preliminary issue; this issue implements a JSON schema.

The next step, is to refine it, for example to address #472 and progress towards and support #507 (a longer journey we can map out with some milestones).

What do you think?

carlhiggs commented:
As there was no further discussion, to continue with progress and close the issue this pull request resolves, I merged this pull request
